### DISCUSSION TITLE
Finding the "thrower" of a molo/incin for inferno_startburn events

### DISCUSSION THREAD

**Discussion by n1c** [saul/demofile#235] [Questions] [answered]
Hey Saul; thanks so much for the awesome lib and all the effort you put in!

I've been working on something to parse demos and get all the he/smoke/flash/molo/incin data. For all the others it's been relatively straight forward to match up the start & end locations (of a throw) by getting the `event userid` on `weapon_fire` and then matching it up when seeing the corresponding event like `flashbang_detonate` or `hegrenade_detonate`.

From what I can tell though there is no `userid` in the `inferno_startburn` event? Have I missed something or is there some special trick to finding out who threw a moly/incin in the startburn event?

Thanks in advance!
Nic.

**Comment by saul** [comment]
Grab the entity of the grenade from the entityid event variable. You can then do `.owner` to get the player that threw the grenade. This may be null if the player threw a grenade then disconnected before it detonated.

**Comment by n1c** [comment]
Thanks for the quick response!

If I have demo `d` and event `e` of type `inferno_startburn`, should I be able to `d.entities.getByHandle(e.entityid)`?

I've tried a few and just seem to be getting nulls, while there is a value for `e.entityid`.

**Comment by saul** [comment]
It’s not a handle - it’s an index. Try `d.entities.entities[e.entityid]`

**Comment by n1c** [comment]
Ah that did the trick, I'd tried `d.entities[e.entityid]` but guess the `entities.entities` eluded me. Thanks!

**Comment by saul** [comment]
No problem @n1c

**Comment by madsboending** [comment]
Hey @n1c 

If you don't mind me asking - how do you "link" the two events _hedetonate_ and _weapon_fire_? 
I'm a complete beginner regarding Node and JS in general, so I apologize for what I fear is a simple question. This is my current approach:

`const demofile = require('demofile');
const fs = require('fs');

fs.readFile('C:/Users/Gamer/Desktop/2018/csgo_demo_parser/match730_003301150011605123236_1345725134_136.dem', function (err, buffer) {
  let demoFile = new demofile.DemoFile();

  demoFile.gameEvents.on('hegrenade_detonate', e => {

  let currentTick = demoFile.currentTick;
  let thrower = demoFile.entities.getByUserId(e.userid);
  let throwerName = thrower ? thrower.name : 'unnamed';
  let eventName = 'hegrenade_detonate';
  let detonatePosition = { x: e.x, y: e.y, z: e.z };
  let userid = e.userid;

  console.log(`${currentTick}, ${userid} ,${eventName}, ${detonatePosition.x}, ${detonatePosition.y}, ${throwerName} `);
});

demoFile.gameEvents.on('weapon_fire', d => {

let currentTick = demoFile.currentTick;
let thrower = demoFile.entities.getByUserId(d.userid);
let userID = d.userid;
let throwerName = thrower ? thrower.name : 'unnamed';
let throwerPositionX = thrower ? thrower.position.x : 'unnamed';
let throwerPositionY = thrower ? thrower.position.y : 'unnamed';

console.log(`${currentTick}, ${userID}, ${d.weapon},  ${throwerPositionX}, ${throwerPositionY}, ${throwerName}`);
});



  demoFile.parse(buffer);
});
`
It has several issues(prints every _weapon_fire_ event and not just HE), but I can't figure out how to link the two events and e.g. just write one line to a file that contains the position of the player when he throws the grenade and the position of the grenade when it detonates.

**Comment by n1c** [comment, marked as answer]
Hey @madsboending; I kinda scrapped this together quickly so hopefully it makes sense: https://gist.github.com/n1c/27dc159ad09740953ea760d7dc1f4ace (also it's in TS hopefully that's still useful)

Maybe Saul can confirm 😉

**Comment by madsboending** [comment]
That helps ALOT! Thank you so much @n1c ! But I'm a bit unsure what you achieve by listening for "round_officially_ended"?

I'm still struggling with figuring out how deep in a round the HEs were thrown, I know I need to substract _currentTime_ of the _round_start_ from the _currentTime_  of the throw, but can't seem to wrap my head around it.

**Comment by n1c** [comment]
The `grenades` hash it used for tracking which grenades have been thrown in the round, and in `round_officially_ended` it's cleared out, this is just to make sure that anything that didn't get caught in the current round doesn't go over and cause bugs. It'd probably be fine to remove that but Safety First! :)

For the time you can probably do something like listen for the `round_start` event, and put that `currentTime` value into a variable. Then in the throw even `roundStartCurrentTime` - `throw tick currentTime` will be what you want,

**Comment by saul** [comment]
Thanks for this @n1c - looks perfect!
